[ENH] ARCH model (arch package) as forecaster

Vasudeva-bit opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://github.com/alan-turing-institute/sktime/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Implements `arch` estimator, Fixes #2173

#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->
A new estimator is added i.e., the `ARCH` model to forecast the volatility. Using the `arch` package, an `arch` estimator is implemented.
#### Does your contribution introduce a new dependency? If yes, which one?

<!--
If your contribution does add a new hard dependency, we may suggest to initially develop your contribution in a separate companion package in https://github.com/sktime/ to keep external dependencies of the core sktime package to a minimum.
-->
Yes, the `arch` package is the soft dependency introduced for implementing the `arch` estimator.
#### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->
Review the implementation of the `arch` model in sktime and also the updated doc, API reference, other files w.r.t to `arch` estimator.
#### Any other comments?
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.
-->
As far as I know, the `arch` package is not supporting the `multivariate` arch models, so need to work on that.! The `update` functionality is implicitly available in the `fit` method of the `arch` model (so, skipped its implementation). Methods such as `predict_intervals` or `predict_quantiles` are yet to be implemented as the `arch` package doesn't support (studying this closely). 

Upcoming updates include:
1. Adding the estimator to online documentation.
2. Add an example notebook to showcase how the `arch` estimator works.
#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions

- I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).

- I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.


<!--
Thanks for contributing!
-->

fkiraly commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,307 @@
+class ARCH(BaseForecaster):
+    """Generalized Autoregressive Conditional Heteroskedasticity (GARCH) model for forecasting 
+    votility in high frequency time series data.
+
```
should say explicitly which package and model therein it interfaces

Vasudeva-bit replied:
Oh okay, understood.

fkiraly replied:
I would say "Directly interfaces ARCH models from python package `arch`"

Vasudeva-bit replied:
Okay, your option is more clear so will update it as "Directly interfaces ARCH models from python package `arch`"

fkiraly commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,307 @@
+    
+    ARCH models are a popular class of volatility models that use observed values of returns or residuals 
+    as volatility shocks. A basic GARCH model is specified as
+        rt = μ+ϵt
```
if you use math, kindly use the math environment in rst.
Also kindly ensure that all symbols are properly defined, if you use math, and clearly linked to the inputs y, X

fkiraly commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,307 @@
+
+    Hyper-parameters
+    ----------------
+    y : {np.ndarray, Series, None}
```
y and x are not hyperparameters

Also, kindly adhere to numpydoc, this should be called "Parameters"

Vasudeva-bit replied:
Yeah, I shouldn't have overlooked this part.

fkiraly commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,307 @@
+
+    _tags = {
+        "scitype:y": "univariate", 
+        "y_inner_mtype": ("np.ndarray", "pd.Series"), 
```
this might have to be a list

Vasudeva-bit replied:
I knew that this should be a list but, I thought the `y_inner_mtype` could be a tuple as it is immutable, so I preferred tuple. Anyway, I can change this.

fkiraly replied:
well, it might work as a tuple, but it´s not the official design so might run into trouble somewhere.

Btw I think you are right, design-wise, tuples would have probably been better.
But interface compliance wise, it needs to be list...

Vasudeva-bit replied:
Got it.

fkiraly commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,307 @@
+        super(ARCH, self).__init__()
+
+
+    def _fit(self, y, X=None, **fit_params):
```
please use the correct signature, `def _fit(self, y, X=None, fh=None):`

Kindly refer to the extension templates

Vasudeva-bit replied:
will correct the signature.

fkiraly commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,307 @@
+
+        Parameters
+        ----------
+        y : guaranteed to be of a type in self.get_tag("y_inner_mtype")
```
replace this by the assumption in this special case. Was that numpy and DataFrame?

Vasudeva-bit replied:
Hmm yeah, will do this as well.

fkiraly replied:
btw, are you certain that the `arch` package  can deal with both kinds of input, i.e., numpy and pandas?

Vasudeva-bit replied:
As specified in the [`arch` documentation](https://arch.readthedocs.io/en/latest/univariate/introduction.html), it can deal with both kinds of input. Anyway, I will verify this before the next PR.

fkiraly replied:
hm, ok!
Then the question would be, does it work natively, or are they doing an internal conversion?
If there's an `arch` internal conversion to one of the two, I think it would be better to keep all conversions in `sktime` and convert only to the "real" internal type. On the other hand, if there's implementations for both formats without conversions within `arch`, then that's the right setting.

Vasudeva-bit replied:
I am not sure with the exact implementation in the `arch` package, I will verify that now.

fkiraly replied:
(if it´s difficult to find out, I don´t mind if you leave it as is, as long as it works)

fkiraly commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,307 @@
+        """
+        from arch import arch_model as _ARCH
+
+        self.y = y
```
you don´t need to write y, X to self. You don´t use it later, do you?

Vasudeva-bit replied:
Actually no, will change this.

Vasudeva-bit replied:
Actually no, I will change that.

fkiraly reviewed (changes requested):
Thanks, great work!

Some change requests:
* the linter is failing, ensure you fix the formatting errors. Look for "precommits" in the developer guide, and/or click on "details" below.
* you should fully comply with the extension template, there are a number of small deviations (but they are blocking since we need interface compatibility)
* the ARCH model is only "interesting" if you also return sigma-t, i.e., the heteroskedastic variance forecasts. It might be easiest to return them using the `predict_var`(iance) interface. This is currently not in the extension template, but is in the base class. For convenience, signature is here:

```python
    def _predict_var(
        self,
        fh=None,
        X=None,
    ):
        """Compute/return variance forecasts.

        private _predict_proba containing the core logic, called from predict_var

        Parameters
        ----------
        fh : guaranteed to be ForecastingHorizon
            The forecasting horizon with the steps ahead to to predict.
        X : optional (default=None)
            guaranteed to be of a type in self.get_tag("X_inner_mtype")
            Exogeneous time series to predict from.

        Returns
        -------
        pred_var : pd.DataFrame
            Column names are exactly those of `y` passed in `fit`/`update`.
                For nameless formats, column index will be a RangeIndex.
            Row index is fh. Entries are variance forecasts, for var in col index.
        """
```

Vasudeva-bit commented:
I shall work on the changes requested, the `predict_var` method. I think I need to do a close review before a PR. Thank you very much @fkiraly.

fkiraly commented:
one more thing, that´s non-blocking: I think `arch` does `update` via its `fit`, so `update` of the sktime needs estimator needs to call the `arch` interface's `fit` with the update setting?

Vasudeva-bit commented:
Yeah, the `arch` interface provides the `update` functionality in the `fit` method, so we can use that by specifying the `update_freq` parameter (like you said).

Vasudeva-bit commented:
Please guide me with this: 
Shall I add an `update` method in the `arch` model to implement update functionality by invoking the `fit` method with the mandatory `update_freq` parameter or is it okay to leave it as an option in the `fit` method only?

fkiraly commented:
> Please guide me with this: 
> Shall I add an `update` method in the `arch` model to implement update functionality by invoking the `fit` method with the mandatory `update_freq` parameter or is it okay to leave it as an option in the `fit` method only?

No, kindly leave the `arch` model unchanged.

The idea is as follows:
* when the `fit` of the `sktime` estimator is called, it should call `arch` `fit`, without update
* when the `update` of the `sktime` estimator is called, it should call update functionality of `arch`, if possible. I have not looked at it, but that would probably be a suitable call to `fit` (of `arch`) with update related parameters?

In particular, no non-conformant args should be added to `sktime` `fit`, any update related functionality should be called from the `sktime.update` interface point.

It may be confusing that `arch` does update in its `fit`, but nevertheless it should be invoked from `sktime` `update` and not `sktime` `fit`.

Does this make sense?

fkiraly commented:
PS: kindly address the linting failures - you can check what needs to be fixed by going on details here:
![image](https://user-images.githubusercontent.com/7985502/160002578-59413c5f-5e7e-4068-9fe0-d5f90c5fd948.png)

Vasudeva-bit commented:
Yeah, I will not change the arch model but add an `update` (which is not currently existing) to invoke the `fit` method of the `arch` model internally with the `update_freq` parameter. 
Btw, I could fix the linting issues, anyway will again crosscheck that after adding methods like `_predict_var` and `update`.

Vasudeva-bit commented:
I will clear the linting issues following the details.

fkiraly commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,319 @@
+
+        super(ARCH, self).__init__()
+
+    def _fit(self, y, X=None, fh=None, **fit_params):
```
the methods should not have parameters `**fit_params` etc, only the specified args

Vasudeva-bit replied:
Oh okay, but the `fit` method of the `arch` model takes some other parameters as well. What can I do about that?

Vasudeva-bit replied:
Thanks for letting me know about linting issues, may I know how I can handle multiple parameters of the `fit` method in the `arch` model. (currently **fit_params is used)

Vasudeva-bit commented:
I have cleared all the linting issues as shown in the details but still, the code couldn't pass the code quality test. I hope you can help me with the approach!

fkiraly commented:
>I have cleared all the linting issues as shown in the details but still, the code couldn't pass the code quality test. I hope you can help me with the approach!

I think you haven't read all the sections, there is content further down in the log, for instance here:

![image](https://user-images.githubusercontent.com/7985502/160278187-fe66ae3f-47c7-45f3-87a2-e486e4344971.png)

There is also more further down from this (not in the screenshot), a green "plus" without anything after it means you need to add a newline.

Vasudeva-bit commented:
Yes, there is something more I need to do, it's completely my mistake.

Vasudeva-bit commented:
Thanks for letting me know about linting issues, may I know how I can handle multiple parameters of the `fit` method in the `arch` model. (currently `**fit_params` is used)

fkiraly commented:
> may I know how I can handle multiple parameters of the fit method in the arch model

If you have more parameters than just the data and the forecasting horizon, you need to do as follows:

* pass the parameter to the constructor
* write it to `self` in the constructor
* in `fit`, read it from `self` and pass it to the `fit` method.

This allows to expose all parameter choices via the scikit-learn compatible interface.

For an example of this, you can look at `smoothing_level` in `ExponentialSmoothing`.

aiwalter commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,405 @@
+        Exogenous regressors.  Ignored if model does not permit exogenous
+        regressors.
+
+    Hyper-parameters
```
there should only be a sectiion ´Parameters´

Vasudeva-bit replied:
Hmm okay, I will put everything under `parameters`.

aiwalter commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,405 @@
+
+    See Also
+    --------
+    Autoregressive Integrated Moving Average (ARIMA) models
```
this should only link class names from `sttime`,

aiwalter commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,405 @@
+
+    See Also
+    --------
+    Autoregressive Integrated Moving Average (ARIMA) models
```
```suggestion
    ARIMA
    AutoARIMA
    StatsForecastAutoARIMA
```

fkiraly replied:
yes, that's the reason why readthedocs is failing. These need to be class names, and the reference section is auto-generated.

Vasudeva-bit replied:
Oh okay, I will change accordingly.

aiwalter commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,405 @@
+    >>> from sktime.forecasting.arch import ARCH
+    >>> y = load_airline()
+    >>> forecaster = ARCH(
+    ...    mean="Constant",
+    ...    lags=0,
+    ...    vol="Garch",
+    ...    p=1,
+    ...    o=0,
+    ...    q=1,
+    ...    power=2.0,
+    ...    dist="Normal",
+    ...    hold_back=None,
+    ...    rescale=None,
+    ...):
```
please make this shorter, maybe even default all params is possible

Vasudeva-bit replied:
Okay, I shall make all the parameters default.

aiwalter commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,405 @@
+        "backcast",
+    )
+
+    def __init__(
```
is there a param equivalent to `sp` from other models, which descibed the seasonal periodicity

Vasudeva-bit replied:
I don't think there is any parameter in the `arch` interface that's equivalent to seasonal periodicity (`sp`).

aiwalter commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,405 @@
+        X : optional (default=None)
+            guaranteed to be of a type in np.ndarray or pd.DataFrame
+            Exogeneous time series to fit to.
+        update_freq : int, optional
+            Frequency of iteration updates.  Output is generated every
+            update_freq iterations. Set to 0 to disable iterative output.
+        disp : {bool, "off", "final"}
+            Either 'final' to print optimization result or 'off' to display
+            nothing. If using a boolean, False is "off" and True is "final"
+        starting_values : np.ndarray, optional
+            Array of starting values to use.  If not provided, starting values
+            are constructed by the model components.
+        cov_type : str, optional
+            Estimation method of parameter covariance.  Supported options are
+            'robust', which does not assume the Information Matrix Equality
+            holds and 'classic' which does.  In the ARCH literature, 'robust'
+            corresponds to Bollerslev-Wooldridge covariance estimator.
+        show_warning : bool, optional
+            Flag indicating whether convergence warnings should be shown
+        first_obs : {int, str, datetime, Timestamp}
+            First observation to use when estimating model
+        last_obs : {int, str, datetime, Timestamp}
+            Last observation to use when estimating model
+        tol : float, optional
+            Tolerance for termination.
+        options : dict, optional
+            Options to pass to 'scipy.optimize.minimize'.  Valid entries
+            include 'ftol', 'eps', 'disp', and 'maxiter'.
+        backcast : {float, np.ndarray}, optional
+            Value to use as backcast. Should be measure '\sigma^2_0'
+            since model-specific non-linear transformations are applied to
+            value before computing the variance recursions.
```
why this here?

aiwalter commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,405 @@
+        y_pred : pd.Series
+            Point predictions
+        """
+        if fh is not None:
+            self.fh = fh
```
I think this is not needed and done in base class, but you might just check that. There is `check_fh()` called I think

Vasudeva-bit replied:
okay, will replace it with the `check_fit` method.

fkiraly replied:
no, kindly do not do anything of that kind at all. Input checks are carried out by the base class.
All the inner methods should implement is model logic.

Vasudeva-bit replied:
Oh, I got what you meant.

fkiraly replied:
what happens if you pass the horizon both in `fit` and `predict`, to `arch`?

Vasudeva-bit replied:
Very sorry, excuse me. Only predict method takes horizon param. The issue is resolved.

fkiraly commented:
> Btw I used **fit_params in the _fit method after looking at the code of adapters such as _pmdarima.py or _fbprophet.py.

Indeed! That looks like an oversight, from legacy (old) code when we hadn't settled yet on the interface patterns.
Have fixed this here: https://github.com/alan-turing-institute/sktime/pull/2343

aiwalter commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,405 @@
+        simulations=None,
+        rng=None,
+        random_state=None,
+        reindex=None,
```
what does `reindex` do? I think we might drop this param if its operating on index

aiwalter reviewed (changes requested):
left few comments

Vasudeva-bit commented:
Thank you @aiwalter. I will do the requested changes.

Vasudeva-bit commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,405 @@
+        fh : guaranteed to be ForecastingHorizon or None, optional (default=None)
+            The forecasting horizon with the steps ahead to to predict.
+            If not passed in _fit, guaranteed to be passed here
+        params : {np.ndarray, Series}
+            Parameters required to forecast. Must be identical in shape to the
+            parameters computed by fitting the model.
+        start : {int, datetime, Timestamp, str}, optional
+            An integer, datetime or str indicating the first observation to
+            produce the forecast for. Datetimes can only be used with pandas
+            inputs that have a datetime index. Strings must be convertible to
+            a date time, such as in '1945-01-01'.
+        align : str, optional
+            Either 'origin' or 'target'. When set of 'origin', the t-th row of
+            forecasts contains the forecasts for t+1, t+2, ..., t+h. When set to
+            'target', the t-th row contains the 1-step ahead forecast from time t-1,
+            the 2 step from time t-2, ..., and the h-step from time t-h. 'target'
+            simplified computing forecast errors since the realization and h-step
+            forecast are aligned.
+        method : {'analytic', 'simulation', 'bootstrap'}
+            Method to use when producing the forecast. The default is analytic. The
+            method only affects the variance forecast generation. Not all volatility
+            models support all methods. In particular, volatility models that do not
+            evolve in squares such as EGARCH or TARCH do not support the 'analytic'
+            method for horizons > 1.
+        simulations : int
+            Number of simulations to run when computing the forecast using either
+            simulation or bootstrap.
+        rng : callable, optional
+            Custom random number generator to use in simulation-based forecasts. Must
+            produce random samples using the syntax rng(size) where size the 2-element
+            tuple (simulations, horizon).
+        random_state : RandomState, optional
+            NumPy RandomState instance to use when method is 'bootstrap'
+        reindex : bool, optional
+            Whether to reindex the forecasts to have the same dimension as the series
+            being forecast. Prior to 4.18 this was the default. As of 4.19 this is now
+            optional. If not provided, a warning is raised about the future change in
+            the default which will occur after September 2021.
```
Do I need to replace these details of parameters (parameters of `_predict()` method) and put them in the constructor or do I need to remove them without replacement?

aiwalter replied:
model related params schould only be exposed in constructor. Where you use them internally is not relevant. I think that docstring was lost here?

fkiraly commented:
> How can I address the failure of test Install and test / test-windows (3.8) (pull_request) ?

Generally, you should be able to run the tests locally, either via `pytest` or `check_estimators`, and then see which tests are failing.
In your case, `_predict` always breaks - you should *not* be trying to write to `fh`. If you read the docstring, all you should do is write model parameters. Do not do any input checks or try to replicate logic around `fh`.

aiwalter commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,405 @@
+        """
+        from arch import arch_model as _ARCH
+
+        self.fh = fh
```
```suggestion
```

Vasudeva-bit commented:
Yes, but I have addressed issues such as side effects to `fh`, and updated the `See Also` section with suggested changes. I will go through the logs once again.

Vasudeva-bit commented:
I didn't push because I thought of addressing the rest of the failed tests like `Install and test / test-windows (3.7) (pull_request)`.

aiwalter commented:
seems the example code in docstring is failing, can you try to run it to see what is wrong there?

Vasudeva-bit commented:
Okay, I will do that.

Vasudeva-bit commented:
I tried testing the example code and there is a weird case when the `params` parameter of the `arch` `forecast` method is `None`, an error is raised. I couldn't understand the exact cause because when I tried testing the `forecast` function directly with the `None` value to the `params` parameter, the code worked perfectly well. I am still trying to resolve it.

fkiraly commented:
I think it´s not just the example code - the estimator `ARCH` is failing a number of tests, check out the logs for what is going wrong.

Vasudeva-bit commented:
Yes, that's right, I understood it from the logs, I will try to fix those.

Vasudeva-bit commented:
I had some exams last week so could not spare enough time working on the arch estimator. I am still working on fixing the issues and I hope I can spend a good amount of time on this task from now. Will reach out to you if I am stuck!

Vasudeva-bit commented:
Hello @fkiraly 
I made the editable install successfully! The error `horizon must be integer >= 1` is because the `fh` is internally converted to class `ForecastingHorizon`. Since `horizon` in the `arch` package is not compatible with class type `ForecastingHorizon` it raised a general error. Using `fh[0]` to get the value for `fh` solves the issue, hope this is a good move. Kindly let me know the alternatives if any.

fkiraly commented:
> Kindly let me know the alternatives if any.

`fh[0]` will only give you the first horizon value, I don´t think that´s a good idea. You want to make predictions at all horizon values.

Vasudeva-bit commented:
Okay, I got it. btw the `arch` `forecast` method is compatible with only int type value for `fh` forecasting horizon. In order to forecast at all values, I think the `forecast` method needs to be invoked multiple times.

fkiraly commented:
>btw the arch forecast method is compatible with only int type value for fh forecasting horizon. 

You can convert back/forth between date if you need to, one of the methods of `ForecastingHorizon` should do this (also see the other estimators)

> In order to forecast at all values, I think the forecast method needs to be invoked multiple times.

Why not just a single time up to the maximum horizon index?

Vasudeva-bit commented:
Yes, forecasting up to the maximum index works. I am also going through methods of class `ForecastingHorizon`.

fkiraly commented:
@Vasudeva-bit, are you still working on this?

Vasudeva-bit commented:
I am working on fixing errors, and hope to resolve them soon

lbventura commented on sktime/forecasting/arch.py:
```diff
@@ -0,0 +1,392 @@
+        -------
+        self : returns an instance of self
+        """
+        from arch import arch_model as _ARCH
```
Should this import occur inside the _fit function?

fkiraly replied:
no, because it is a soft dependency.

All soft dependency imports should be isolated in a way like this, so importing the module does not error out under the minimal dependency set (which excludes soft dependencies).

All estimators should also be aware of the soft dependencies in them, via the `python_dependencies` tag.
